src/Workspaces/Remote/Core/Services/AssetService.cs

```diff
@@ -172,6 +239,88 @@ public Entry(object @object)
                 Object = @object;
             }
         }
+
+        private struct ChecksumSynchronizer
+        {
+            private readonly AssetService _assetService;
+
+            public ChecksumSynchronizer(AssetService assetService)
+            {
+                _assetService = assetService;
+            }
+
+            public async Task SynchronizeAsync(Checksum solutionChecksum, CancellationToken cancellationToken)
+            {
+                // get children of solution checksum object at once
+                var solutionChecksums = new HashSet<Checksum>();
+                var solutionChecksumObject = await _assetService.GetAssetAsync<SolutionChecksumObject>(solutionChecksum, cancellationToken).ConfigureAwait(false);
+
+                AppendIfNeeded(solutionChecksums, solutionChecksumObject.Info);
+                AppendIfNeeded(solutionChecksums, solutionChecksumObject.Projects);
+
+                await _assetService.SynchronizeAssetsAsync(solutionChecksums, cancellationToken).ConfigureAwait(false);
+
+                // get children of project checksum objects at once
+                var projectChecksums = new HashSet<Checksum>();
```

Comment:
I made Solution/Project checksum object to just have its direct children's information so that I don't need to send over too many information unnecessary vs making communication too chatty.

so I am getting each chunk at once, solution and its direct children, projects and their direct children and documents and their direct children.

I can't know direct children of projects until I get direct children (projects) of solution. same for document.

initially, I made one checksum object to hold onto whole hierarchy, and that made a single character change to send over 25k (number of checksums) * 20 bytes (size of checksum) for roslyn-internal each time since hierarchy checksum changed.

now each level has only its direct children. so about 100~200 (number of checksums) x 20 bytes (size of checksum)

so most likely with change, it will re-sync solution (2-4k), project (2-4k) and source text that has changed.

didn't want to go down more than that to make it not to be too chatty.

some of these might need to be more granular than now to support open file IDE features like intellisense, but right now all usages is something that assume much longer delay so probably fine.

also, time it takes to send delay is still around 1-2ms so, shouldn't be too much issue.